LinuxAdapter.cs

```diff
@@ -0,0 +1,120 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using System;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Authentication.Negotiate
+{
+    public class LinuxAdapter
+    {
+        private readonly string _distinguishedName;
+        private readonly LdapConnection _connection;
+
+        public LinuxAdapter(string machineAccount, string machinePassword)
+        {
+            if (string.IsNullOrEmpty(machineAccount))
+            {
+                throw new ArgumentNullException(nameof(machineAccount));
+            }
+            if (string.IsNullOrEmpty(machinePassword))
+            {
+                throw new ArgumentNullException(nameof(machinePassword));
+            }
+
+            var domain = machineAccount.Substring(machineAccount.IndexOf('@') + 1);
+            _distinguishedName = domain.Split('.').Select(name => $"dc={name}").Aggregate((a, b) => $"{a},{b}");
+
+            LdapDirectoryIdentifier di = new LdapDirectoryIdentifier(server: domain, fullyQualifiedDnsHostName: true, connectionless: false);
+            NetworkCredential credential = new NetworkCredential(machineAccount, machinePassword);
+            _connection = new LdapConnection(di, credential);
+            _connection.SessionOptions.ProtocolVersion = 3; //Setting LDAP Protocol to latest version
+            _connection.Bind(); // This line actually makes the connection.
+            _connection.Timeout = TimeSpan.FromMinutes(1);
```

Comment:
I might be misunderstanding but it seems to me that the Timeout here is used for subsequent SendRequest calls. Does it affect bind @joperezr?

Reply:
Timeout property on the connection sets the default timeout of all operations, includding Bind and SendRequest. SendRequest has an overload where you can manually pass in a timeout in case you want to use something other than the default for the connection.